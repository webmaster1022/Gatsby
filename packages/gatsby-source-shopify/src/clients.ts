import fetch, { Response } from "node-fetch"
import { HttpError } from "./errors"

const MAX_BACKOFF_MILLISECONDS = 60000

// TODO Update logic for handling errors (500 specifically) and update the
// "hacky" code in createRestClient

export function createGraphqlClient(
  options: IShopifyPluginOptions
): IGraphQLClient {
  const url = `https://${options.storeUrl}/admin/api/2021-07/graphql.json`

  async function graphqlFetch<T>(
    query: string,
    variables?: Record<string, unknown>,
    retries = 0
  ): Promise<T> {
    const response = await fetch(url, {
      method: `POST`,
      headers: {
        "Content-Type": `application/json`,
        "X-Shopify-Access-Token": options.password,
      },
      body: JSON.stringify({
        query,
        variables,
      }),
    })

    if (!response.ok) {
      const waitTime = 2 ** (retries + 1) + 500
      if (response.status >= 500 && waitTime < MAX_BACKOFF_MILLISECONDS) {
        await new Promise(resolve => setTimeout(resolve, waitTime))
        return graphqlFetch(query, variables, retries + 1)
      }

      throw new HttpError(response)
    }

    const json = await response.json()
    return json.data as T
  }

  return { request: graphqlFetch }
}

export function createRestClient(options: IShopifyPluginOptions): IRestClient {
  const baseUrl = `https://${options.storeUrl}/admin/api/2021-01`

  async function shopifyFetch(
    path: string,
    fetchOptions = {
      headers: {
        "X-Shopify-Access-Token": options.password,
      },
    },
    retries = 3
  ): Promise<Response> {
    /**
     * This is kind of a hack, but...
     *
     * We do this because although callers will use a relative path,
     * some responses might have pagination links that get fed back
     * to shopifyFetch to retrieve the next page. Those links are
     * absolute URLs so we account for both, but not in a very robust
     * fashion.
     */
    const url = path.includes(options.storeUrl) ? path : `${baseUrl}${path}`

    const resp = await fetch(url, fetchOptions)

    if (!resp.ok && retries > 0 && resp.status === 429) {
      // rate limit
      const retryAfter = parseFloat(resp.headers.get(`Retry-After`) || ``)
      await new Promise(resolve => setTimeout(resolve, retryAfter))
      return shopifyFetch(path, fetchOptions, retries - 1)
    }

    return resp
  }

  return {
    request: async (path: string): Promise<Response> => shopifyFetch(path),
  }
}
