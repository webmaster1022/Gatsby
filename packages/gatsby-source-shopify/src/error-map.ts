import { shiftLeft } from "shift-left"
import { pluginErrorCodes as errorCodes } from "./errors"

const getErrorText = (context: IErrorContext): string => context.sourceMessage

export const ERROR_MAP: IErrorMap = {
  [errorCodes.bulkOperationFailed]: {
    text: getErrorText,
    level: `ERROR`,
    category: `USER`,
  },
  [errorCodes.apiConflict]: {
    text: (): string => shiftLeft`
    Your operation was canceled. You might have another production site for this Shopify store.

    Shopify only allows one bulk operation at a time for a given shop, so we recommend that you
    avoid having two production sites that point to the same Shopify store.

    If the duplication is intentional, please wait for the other operation to finish before trying
    again. Otherwise, consider deleting the other site or pointing it to a test store instead.
  `,
    level: `ERROR`,
    category: `USER`,
  },
  /**
   * If we don't know what it is, we haven't done our due
   * diligence to handle it explicitly. That means it's our
   * fault, so THIRD_PARTY indicates us, the plugin authors.
   */
  [errorCodes.unknownSourcingFailure]: {
    text: getErrorText,
    level: `ERROR`,
    category: `THIRD_PARTY`,
  },
  [errorCodes.unknownApiError]: {
    text: getErrorText,
    level: `ERROR`,
    category: `THIRD_PARTY`,
  },
}
