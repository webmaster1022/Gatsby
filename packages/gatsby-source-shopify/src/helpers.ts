import { NodePluginArgs, SourceNodesArgs } from "gatsby"
import { createRemoteFileNode } from "gatsby-source-filesystem"

import { shopifyTypes } from "./shopify-types"

const pattern = /^gid:\/\/shopify\/(\w+)\/(.+)$/

export function createNodeId(
  shopifyId: string,
  gatsbyApi: NodePluginArgs,
  { typePrefix }: IShopifyPluginOptions
): string {
  return gatsbyApi.createNodeId(`${typePrefix}${shopifyId}`)
}

export function getPluginStatus(
  gatsbyApi: NodePluginArgs
): Record<string, never> {
  return (
    gatsbyApi.store.getState().status.plugins?.[`gatsby-source-shopify`] || {}
  )
}

export function getLastBuildTime(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions
): Date | undefined {
  const { typePrefix } = pluginOptions
  const status = getPluginStatus(gatsbyApi)

  return (
    status[`lastBuildTime${typePrefix}`] &&
    new Date(status[`lastBuildTime${typePrefix}`])
  )
}

export function setLastBuildTime(
  gatsbyApi: NodePluginArgs,
  pluginOptions: IShopifyPluginOptions,
  currentBuildTime: number
): void {
  const { typePrefix } = pluginOptions
  const status = getPluginStatus(gatsbyApi)

  gatsbyApi.actions.setPluginStatus({
    ...status,
    [`lastBuildTime${typePrefix}`]: currentBuildTime,
  })
}

export function isPriorityBuild(pluginOptions: IShopifyPluginOptions): boolean {
  const { CI, GATSBY_CLOUD, GATSBY_IS_PR_BUILD, NETLIFY, CONTEXT } = process.env

  const isGatsbyCloudPriorityBuild =
    CI === `true` && GATSBY_CLOUD === `true` && GATSBY_IS_PR_BUILD !== `true`

  const isNetlifyPriorityBuild =
    CI === `true` && NETLIFY === `true` && CONTEXT === `production`

  return pluginOptions.prioritize !== undefined
    ? pluginOptions.prioritize
    : isGatsbyCloudPriorityBuild || isNetlifyPriorityBuild
}

export function isShopifyId(shopifyId: string): boolean {
  return pattern.test(shopifyId)
}

export function parseShopifyId(shopifyId: string): Array<string> {
  return shopifyId.match(pattern) || []
}

export function decorateBulkObject(input: unknown): unknown {
  if (input && typeof input === `object`) {
    if (Array.isArray(input)) {
      return input.map(decorateBulkObject)
    }

    const obj: Record<string, unknown> = { ...input }

    for (const key of Object.keys(obj)) {
      obj[key] = decorateBulkObject(obj[key])
    }

    // We must convert ID to ShopifyID so that it doesn't collide with Gatsby's internal ID
    if (obj.id) {
      obj.shopifyId = obj.id
      delete obj.id
    }

    return obj
  }

  return input
}

export async function processShopifyImages(
  { actions: { createNode }, createNodeId, cache }: SourceNodesArgs,
  node: IShopifyNode
): Promise<void> {
  const type = parseShopifyId(node.shopifyId)[1]
  const imageFields = shopifyTypes[type].imageFields

  if (imageFields) {
    for (const fieldPath of imageFields) {
      const image = fieldPath
        .split(`.`)
        .reduce((acc, value) => acc[value] as never, node) as
        | IShopifyImage
        | undefined

      if (image && parseImageExtension(image.originalSrc) !== `gif`) {
        const fileNode = await createRemoteFileNode({
          url: image.originalSrc,
          cache,
          createNode,
          createNodeId,
          parentNodeId: node.id,
        })

        image.localFile___NODE = fileNode.id
      }
    }
  }
}

export function parseImageExtension(url: string): string {
  const basename = url.split(`?`)[0]
  const dot = basename.lastIndexOf(`.`)

  if (dot !== -1) {
    return basename.slice(dot + 1)
  } else {
    throw new Error(
      `Could not parse file extension from Shopify image URL: ${url}`
    )
  }
}
