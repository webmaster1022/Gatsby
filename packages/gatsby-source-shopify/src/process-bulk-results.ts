import { SourceNodesArgs } from "gatsby"

import { shopifyTypes } from "./shopify-types"
import {
  isShopifyId,
  createNodeId,
  parseShopifyId,
  decorateBulkObject,
  processShopifyImages,
} from "./helpers"

interface IDecoratedResultChildren {
  [key: string]: Array<string | IDecoratedResult>
}

function isNode(result: IDecoratedResult): boolean {
  const keys = Object.keys(result)
  return Boolean(
    result.shopifyId &&
      !(keys.length === 2 && result.__parentId && result.shopifyId)
  )
}

export async function processBulkResults(
  gatsbyApi: SourceNodesArgs,
  pluginOptions: IShopifyPluginOptions,
  results: BulkResults
): Promise<number> {
  let nodeCount = 0
  const promises: Array<Promise<void>> = []
  const children: IDecoratedResultChildren = {}

  for (let i = results.length - 1; i >= 0; i -= 1) {
    const result = decorateBulkObject(results[i]) as IDecoratedResult
    /**
     * @todo detect the following different as JSON.stringify order is not deterministic
     */
    const resultIsNode = isNode(result)

    if (resultIsNode) {
      const type = parseShopifyId(result.shopifyId)[1]
      const imageFields = shopifyTypes[type].imageFields
      const referenceFields = shopifyTypes[type].referenceFields

      // Only increment for parent nodes
      if (!result.__parentId) nodeCount++

      // Attach reference fields
      if (referenceFields) {
        for (const referenceField of referenceFields) {
          result[referenceField] = []
        }
      }

      // Attach children references / objects
      if (children[result.shopifyId]) {
        for (const child of children[result.shopifyId]) {
          const childType =
            typeof child === `string`
              ? parseShopifyId(child)[1]
              : (child.__typename as string)
          const field = shopifyTypes[childType].key

          if (field) {
            result[field] = [
              typeof child === `string`
                ? createNodeId(child, gatsbyApi, pluginOptions)
                : child,
              ...((result[field] || []) as Array<unknown>),
            ]
          } else {
            throw new Error(
              `There is no shopifyType for child type: ${childType}`
            )
          }
        }
      }

      if (!isShopifyId(result.shopifyId)) {
        throw new Error(
          `Expected an ID in the format gid://shopify/<typename>/<id>`
        )
      }

      const node = {
        ...result,
        id: createNodeId(result.shopifyId, gatsbyApi, pluginOptions),
        internal: {
          type: `${pluginOptions.typePrefix}Shopify${type}`,
          contentDigest: gatsbyApi.createContentDigest(result),
        },
      } as IShopifyNode

      if (pluginOptions.downloadImages && imageFields) {
        promises.push(
          processShopifyImages(gatsbyApi, node).then(() =>
            gatsbyApi.actions.createNode(node)
          )
        )
      } else {
        gatsbyApi.actions.createNode(node)
      }
    }

    // If the object is a child store the reference
    if (result.__parentId) {
      children[result.__parentId] = [
        ...(children[result.__parentId] || []),
        result.shopifyId || result,
      ]
    }
  }

  await Promise.all(promises)
  return nodeCount
}
