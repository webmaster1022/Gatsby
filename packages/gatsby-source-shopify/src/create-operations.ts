import { SourceNodesArgs } from "gatsby"

import { createGraphqlClient } from "./clients"
import { OperationError } from "./errors"

import { ProductsQuery } from "./query-builders/products-query"
import { ProductVariantsQuery } from "./query-builders/product-variants-query"
import { CollectionsQuery } from "./query-builders/collections-query"
import { OrdersQuery } from "./query-builders/orders-query"
import { LocationsQuery } from "./query-builders/locations-query"

import {
  OPERATION_STATUS_QUERY,
  OPERATION_BY_ID,
  CANCEL_OPERATION,
} from "./static-queries"

interface IOperations {
  productsOperation: IShopifyBulkOperation
  productVariantsOperation: IShopifyBulkOperation
  ordersOperation: IShopifyBulkOperation
  collectionsOperation: IShopifyBulkOperation
  locationsOperation: IShopifyBulkOperation
  cancelOperationInProgress: () => Promise<void>
  cancelOperation: (id: string) => Promise<IBulkOperationCancelResponse>
  finishLastOperation: () => Promise<void>
  completedOperation: (
    operationId: string,
    interval?: number
  ) => Promise<{ node: IBulkOperationNode }>
}

const finishedStatuses = [`COMPLETED`, `FAILED`, `CANCELED`, `EXPIRED`]
const failedStatuses = [`FAILED`, `CANCELED`]

export function createOperationObject(
  graphqlClient: IGraphQLClient,
  operationQuery: string,
  name: string
): IShopifyBulkOperation {
  return {
    execute: (): Promise<IBulkOperationRunQueryResponse> =>
      graphqlClient.request<IBulkOperationRunQueryResponse>(operationQuery),
    name,
  }
}

export function createOperations(
  gatsbyApi: SourceNodesArgs,
  pluginOptions: IShopifyPluginOptions,
  lastBuildTime?: Date
): IOperations {
  const graphqlClient = createGraphqlClient(pluginOptions)

  function currentOperation(): Promise<ICurrentBulkOperationResponse> {
    return graphqlClient.request(OPERATION_STATUS_QUERY)
  }

  async function finishLastOperation(): Promise<void> {
    let { currentBulkOperation } = await currentOperation()
    if (currentBulkOperation && currentBulkOperation.id) {
      if (!finishedStatuses.includes(currentBulkOperation.status)) {
        const timer = gatsbyApi.reporter.activityTimer(
          `Waiting for operation ${currentBulkOperation.id} : ${currentBulkOperation.status}`
        )
        timer.start()

        while (!finishedStatuses.includes(currentBulkOperation.status)) {
          await new Promise(resolve => setTimeout(resolve, 1000))
          currentBulkOperation = (await currentOperation()).currentBulkOperation
          timer.setStatus(
            `Polling operation ${currentBulkOperation.id} : ${currentBulkOperation.status}`
          )
        }

        timer.end()
      }
    }
  }

  async function cancelOperation(
    id: string
  ): Promise<IBulkOperationCancelResponse> {
    return graphqlClient.request<IBulkOperationCancelResponse>(
      CANCEL_OPERATION,
      {
        id,
      }
    )
  }

  async function cancelOperationInProgress(): Promise<void> {
    let { currentBulkOperation: bulkOperation } = await currentOperation()

    if (!bulkOperation || finishedStatuses.includes(bulkOperation.status)) {
      return
    }

    const cancelTimer = gatsbyApi.reporter.activityTimer(
      `Cancelling previous operation: ${bulkOperation.id}`
    )

    cancelTimer.start()

    if (bulkOperation.status === `RUNNING`) {
      const { bulkOperationCancel } = await cancelOperation(bulkOperation.id)

      bulkOperation = bulkOperationCancel.bulkOperation

      while (bulkOperation.status !== `CANCELED`) {
        await new Promise(resolve => setTimeout(resolve, 100))
        const currentOp = await currentOperation()
        bulkOperation = currentOp.currentBulkOperation
        cancelTimer.setStatus(
          `Waiting for operation to cancel: ${bulkOperation.id}`
        )
      }
      cancelTimer.setStatus(`Cancelled operation: ${bulkOperation.id}`)
    } else {
      /**
       * Just because it's not running doesn't mean it's done. For
       * example, it could be CANCELING. We still have to wait for it
       * to be officially finished before we start a new one.
       */
      while (!finishedStatuses.includes(bulkOperation.status)) {
        await new Promise(resolve => setTimeout(resolve, 100))
        bulkOperation = (await currentOperation()).currentBulkOperation
        cancelTimer.setStatus(
          `Waiting for operation to complete: ${bulkOperation.id}`
        )
      }
      cancelTimer.setStatus(`Completed operation: ${bulkOperation.id}`)
    }

    cancelTimer.end()
  }

  /* Maybe the interval should be adjustable, because users
   * with larger data sets could easily wait longer. We could
   * perhaps detect that the interval being used is too small
   * based on returned object counts and iteration counts, and
   * surface feedback to the user suggesting that they increase
   * the interval.
   */
  async function completedOperation(
    operationId: string,
    interval = 1000
  ): Promise<{ node: IBulkOperationNode }> {
    let operation = await graphqlClient.request<{
      node: IBulkOperationNode
    }>(OPERATION_BY_ID, {
      id: operationId,
    })

    let waitForOperation = true

    while (waitForOperation) {
      if (failedStatuses.includes(operation.node.status)) {
        waitForOperation = false
        throw new OperationError(operation.node)
      }

      if (operation.node.status === `COMPLETED`) {
        waitForOperation = false
        return operation
      }

      await new Promise(resolve => setTimeout(resolve, interval))

      operation = await graphqlClient.request<{
        node: IBulkOperationNode
      }>(OPERATION_BY_ID, {
        id: operationId,
      })
    }

    throw new Error(`It should never reach this error`)
  }

  return {
    productsOperation: exports.createOperationObject(
      graphqlClient,
      new ProductsQuery(pluginOptions).query(lastBuildTime),
      `products`
    ),

    productVariantsOperation: exports.createOperationObject(
      graphqlClient,
      new ProductVariantsQuery(pluginOptions).query(lastBuildTime),
      `variants`
    ),

    ordersOperation: exports.createOperationObject(
      graphqlClient,
      new OrdersQuery(pluginOptions).query(lastBuildTime),
      `orders`
    ),

    collectionsOperation: exports.createOperationObject(
      graphqlClient,
      new CollectionsQuery(pluginOptions).query(lastBuildTime),
      `collections`
    ),

    locationsOperation: exports.createOperationObject(
      graphqlClient,
      new LocationsQuery(pluginOptions).query(lastBuildTime),
      `locations`
    ),

    cancelOperationInProgress,
    cancelOperation,
    finishLastOperation,
    completedOperation,
  }
}
