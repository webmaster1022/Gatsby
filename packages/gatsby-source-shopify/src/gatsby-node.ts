import { NodePluginArgs } from "gatsby"

import { ERROR_MAP } from "./error-map"

export { sourceNodes } from "./source-nodes"
export { createResolvers } from "./create-resolvers"
export { pluginOptionsSchema } from "./plugin-options-schema"
export { createSchemaCustomization } from "./create-schema-customization"

const initializePlugin = ({ reporter }: NodePluginArgs): void => {
  reporter.setErrorMap(ERROR_MAP)
}

let coreSupportsOnPluginInit: `unstable` | `stable` | undefined

try {
  const { isGatsbyNodeLifecycleSupported } = require(`gatsby-plugin-utils`)
  if (isGatsbyNodeLifecycleSupported(`onPluginInit`)) {
    coreSupportsOnPluginInit = `stable`
  } else if (isGatsbyNodeLifecycleSupported(`unstable_onPluginInit`)) {
    coreSupportsOnPluginInit = `unstable`
  }
} catch (e) {
  console.error(`Could not check if Gatsby supports onPluginInit lifecycle`)
}

if (coreSupportsOnPluginInit === `unstable`) {
  // need to conditionally export otherwise it throws an error for older versions
  exports.unstable_onPluginInit = initializePlugin
} else if (coreSupportsOnPluginInit === `stable`) {
  exports.onPluginInit = initializePlugin
} else {
  exports.onPreInit = initializePlugin
}
