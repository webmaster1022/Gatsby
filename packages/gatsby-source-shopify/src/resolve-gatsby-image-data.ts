import { readFileSync } from "fs"
import { GatsbyCache } from "gatsby"
import { IGatsbyImageFieldArgs } from "gatsby-plugin-image/graphql-utils"
import { fetchRemoteFile } from "gatsby-core-utils/fetch-remote-file"
import {
  generateImageData,
  getLowResolutionImageURL,
  IGatsbyImageData,
  IGatsbyImageHelperArgs,
  IImage,
  ImageFormat,
} from "gatsby-plugin-image"

import { urlBuilder } from "./get-shopify-image"
import { parseImageExtension } from "./helpers"

type IImageWithPlaceholder = IImage & {
  placeholder: string
}

async function getImageBase64({
  imageAddress,
  directory,
  contentDigest,
}: {
  imageAddress: string
  directory: string
  contentDigest: string
}): Promise<string> {
  // Downloads file to the site cache and returns the file path for the given image (this is a path on the host system, not a URL)
  const filePath = await fetchRemoteFile({
    url: imageAddress,
    directory,
    cacheKey: contentDigest,
  })
  const buffer = readFileSync(filePath)
  return buffer.toString(`base64`)
}

/**
 * Download and generate a low-resolution placeholder
 *
 * @param lowResImageFile
 */
function getBase64DataURI({ imageBase64 }: { imageBase64: string }): string {
  return `data:image/png;base64,${imageBase64}`
}

export function makeResolveGatsbyImageData(cache: GatsbyCache) {
  return async function resolveGatsbyImageData(
    image: IShopifyImage,
    {
      formats = [`auto`],
      layout = `constrained`,
      ...remainingOptions
    }: IGatsbyImageFieldArgs
  ): Promise<IGatsbyImageData | null> {
    // Sharp cannot optimize GIFs so we must return null in that case
    const ext = parseImageExtension(image.originalSrc)
    if (ext === `gif`) return null

    let placeholderURL

    const generateImageSource: IGatsbyImageHelperArgs["generateImageSource"] = (
      filename,
      width,
      height,
      toFormat
    ): IImageWithPlaceholder => {
      return {
        width,
        height,
        placeholder: ``,
        format: toFormat,
        src: urlBuilder({
          width,
          height,
          baseUrl: filename,
          format: toFormat,
          options: {},
        }),
      }
    }

    const sourceMetadata = {
      width: image.width,
      height: image.height,
      format: ext as ImageFormat,
    }

    if (remainingOptions && remainingOptions.placeholder === `blurred`) {
      // This function returns the URL for a 20px-wide image, to use as a blurred placeholder
      const lowResImageURL = getLowResolutionImageURL({
        ...remainingOptions,
        formats,
        layout,
        sourceMetadata,
        pluginName: `gatsby-source-shopify`,
        filename: image.originalSrc,
        generateImageSource,
      })

      const imageBase64 = await getImageBase64({
        imageAddress: lowResImageURL,
        directory: cache.directory as string,
        contentDigest: image.internal.contentDigest,
      })

      placeholderURL = getBase64DataURI({ imageBase64 })
    }
    return generateImageData({
      ...remainingOptions,
      placeholderURL,
      formats,
      layout,
      sourceMetadata,
      pluginName: `gatsby-source-shopify`,
      filename: image.originalSrc,
      generateImageSource,
    })
  }
}
