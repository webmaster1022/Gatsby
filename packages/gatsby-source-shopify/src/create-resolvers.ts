import { CreateResolversArgs } from "gatsby"
import { getGatsbyImageFieldConfig } from "gatsby-plugin-image/graphql-utils"

import { shopifyTypes } from "./shopify-types"
import { makeResolveGatsbyImageData } from "./resolve-gatsby-image-data"

export function createResolvers(
  gatsbyApi: CreateResolversArgs,
  pluginOptions: IShopifyPluginOptions
): void {
  const { createResolvers, cache } = gatsbyApi
  const {
    downloadImages,
    typePrefix,
    shopifyConnections: connections,
  } = pluginOptions

  if (!downloadImages) {
    createResolvers({
      [`${typePrefix}ShopifyImage`]: {
        gatsbyImageData: {
          ...getGatsbyImageFieldConfig(makeResolveGatsbyImageData(cache)),
          type: `JSON`, // Because gatsbyImageData will be undefined for GIFs it must be optional
        },
      },
    })
  }

  // Attach the metafield resolver to all types with a metafields field
  for (const [type, value] of Object.entries(shopifyTypes)) {
    if (value.coupledNodeFields?.includes(`metafields___NODE`)) {
      // Only include the resolver if the type is included in the build
      if (!value.optionalKey || connections.includes(value.optionalKey)) {
        createResolvers({
          [`${typePrefix}Shopify${type}`]: {
            metafield: {
              resolve: async (
                source: IShopifyNode,
                args: { key: string; namespace: string },
                context: {
                  nodeModel: {
                    findOne: (object: unknown) => IShopifyNode | undefined
                  }
                }
              ) =>
                context.nodeModel.findOne({
                  type: `${typePrefix}ShopifyMetafield`,
                  query: {
                    filter: {
                      key: { eq: args.key },
                      namespace: { eq: args.namespace },
                      id: { in: source.metafields___NODE },
                    },
                  },
                }),
            },
          },
        })
      }
    }
  }
}
