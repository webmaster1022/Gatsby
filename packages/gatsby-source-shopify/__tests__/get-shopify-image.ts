import { getShopifyImage } from "../src/get-shopify-image"

const image = {
  originalSrc: `https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero.jpg?v=1460125603`,
  width: 2048,
  height: 1535,
}

describe(`getShopifyImage`, () => {
  it(`generates an imagedata object`, () => {
    const data = getShopifyImage({ image, layout: `fullWidth` })
    expect(data).toMatchSnapshot()
  })
})
