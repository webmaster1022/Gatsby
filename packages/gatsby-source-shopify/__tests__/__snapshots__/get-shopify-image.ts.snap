// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getShopifyImage generates an imagedata object 1`] = `
Object {
  "backgroundColor": undefined,
  "height": 0.74951171875,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_2048x1535_crop_center.jpg?v=1460125603",
      "srcSet": "https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_320x240_crop_center.jpg?v=1460125603 320w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_654x490_crop_center.jpg?v=1460125603 654w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_768x576_crop_center.jpg?v=1460125603 768w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_1024x768_crop_center.jpg?v=1460125603 1024w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_1366x1024_crop_center.jpg?v=1460125603 1366w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_1600x1199_crop_center.jpg?v=1460125603 1600w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_1920x1439_crop_center.jpg?v=1460125603 1920w,
https://cdn.shopify.com/s/files/1/0854/5382/products/front-hero_2048x1535_crop_center.jpg?v=1460125603 2048w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;
